### ISSUE TITLE
Server completely un-joinable after restart, corrupted .explored files?

### ISSUE THREAD

**Issue by brahaney** [Mydayyy/Valheim-ServerSideMap#18]
First off-- love the mod concept. Until a restart it works great! My friends and I were loving the sharing. If theres any way I can help to further debug this or anything please let me know. I'm a Software Engineer in Test so, testing/debugging is my jam. Is there a way to activate like a debug mode for BepInEx or just this mod, to get some logging from the events I describe below? Currently nothing at all happens in the logs.

## The issue
This issue is consistently replicable in my environment. I have not yet tried isolating just this mod to replicate but if I have time to do so I will post my findings.

My server is scheduled to do check for updates nightly, which currently results in a restart. Each morning this results in being completely unable to reconnect to the server. There are no errors in the server log or client log-- when connecting, instead of prompting for my password, I get the following two screens:

![image](https://user-images.githubusercontent.com/4593316/110243499-a9419900-7f28-11eb-8ff1-4b7adab55e37.png)
![image](https://user-images.githubusercontent.com/4593316/110243508-b2cb0100-7f28-11eb-9a6f-d6dec51257dc.png)

The connection issue applies to all users, not only users with the mod installed. One of our players is still running completely vanilla client and cannot connect either once this issue begins.

I've deduced this to be an issue with SSM. It's the only mod I had added since the crashes started, and I can resolve the issue without removing the mod from the server or by simply deleting or moving `Wootheim2.mod.serversidemap.explored` and restarting.

This issue has replicated each time my server has restarted to check for updates, and can be replicated in my environment below with a simple `docker-compose restart` command.

I have also attached two of the "corrupted" (I guess?) `.explored` files, not sure if they're helpful.
[wootheim-explored-files.zip](https://github.com/Mydayyy/Valheim-ServerSideMap/files/6097334/wootheim-explored-files.zip)

## Environment
latest version of Valheim dedicated server, linux (containerized in docker), with only a handful of mods:
* BepInEx `5.4.8.0`
* [ValheimPlus](https://github.com/valheimPlus/ValheimPlus/releases/tag/0.9.4) `0.9.4`
* [CraftFromContainers](https://www.nexusmods.com/valheim/mods/40?tab=files) `1.1.2`run by this [docker image](https://github.com/mbround18/valheim-docker)
* ServerSideMap (`rc1-1.2.0` & `1.2.0`) (default config, no pin sharing)

I also attached my zipped config directory in case anything in there helps. 
[wootheim-config.zip](https://github.com/Mydayyy/Valheim-ServerSideMap/files/6097324/wootheim-config.zip)

I run the entire thing in docker, using [mbround18/valheim](https://github.com/mbround18/valheim-docker) docker image and the following `docker-compose.yml`

```yml
version: "3"
services:
  valheim:
    image: mbround18/valheim:latest
    ports:
      - 2456:2456/udp
      - 2457:2467/udp
      - 2458:2458/udp
    environment:
      - PORT=2456
      - NAME="Wootheim"
      - WORLD="Wootheim2"
      - PASSWORD="coolkids!"
      - TZ=America/Chicago
      - PUBLIC=1
      - AUTO_UPDATE=1
      - AUTO_UPDATE_SCHEDULE="0 1 * * *"
      - AUTO_BACKUP=1
      - AUTO_BACKUP_SCHEDULE="*/15 * * * *"
      - AUTO_BACKUP_REMOVE_OLD=1
      - AUTO_BACKUP_DAYS_TO_LIVE=3
      - AUTO_BACKUP_ON_UPDATE=1
      - AUTO_BACKUP_ON_SHUTDOWN=1
      - UPDATE_ON_STARTUP=0
    volumes:
      - /home/valheim/saves:/home/steam/.config/unity3d/IronGate/Valheim
      - /home/valheim/server:/home/steam/valheim
      - /home/valheim/backups:/home/steam/backups
```

**Comment by Mydayyy** [comment]
Oooh thats certainly interesting. So far I had another user run into the same problem and I thought it was an issue with the used docker image (which was pterodactyl), but I couldnt for the life of me reproduce it, unfortunately he turned inactive.

I'll take a look at everything and try to reproduce it, I hope I can catch that slippery bug this time. The bug report is certainly extremely useful, thank you :)

Cheers 
Mydayyy

**Comment by weebs** [comment]
Looks like this is why I had to disable the mod on my server. I assumed it was due to the newest update being incompatible with the patch, but after removing the .explored file, I'm able to use the mod again.

Something I noticed about the issue is the valheim server process RAM usage would quickly explode to 100% and essentially stall out my 4GB VPS instance. I wonder if my machine had more RAM it would be able to get the server up and running?

I'm running an ubuntu based VPS with the dedicated server setup through LGSM (https://linuxgsm.com/lgsm/vhserver/)

**Comment by Mydayyy** [comment]
He, that could certainly be it, and I can reproduce that. I wonder where I screwed up, wow. I'll take in 2 hours, but that should definitely not be happening but thats atleast something thats somehow easy to find/track

**Comment by brahaney** [comment]
Ok so it seems like its related to how much data is in the shared map-- Me and my friends all did a bunch of exploring late last night, as well as synced all of our maps with the server for the first time. After the reboot that night, my DigitalOcean droplet (2CPU/4GB RAM) was totally unresponsive, to the point where I couldn't even connect an SSH session. But since then I reproduced the issue a few times without any performance issues by starting the server, connecting to sync my map, and then restarting the server.

**Comment by Mydayyy** [comment]
Greetings,

I put out a rc-1.2.1 which should fix that issue: https://github.com/Mydayyy/Valheim-ServerSideMap/releases/download/rc1-1.2.1/ServerSideMap.zip

What went wrong?
I was using `var data = reader.ReadBytes(int.MaxValue);` to read all bytes in the exploration file at once, and has stated by the documentation: Passing int.maxValue might return less bytes if an end of stream is reached. What it didn't tell me is that, .net will still allocate a freaking 4GB array of data, even if the file size is less than an int.MaxValue.

Found that out via https://stackoverflow.com/a/8613300

If you can spare a minute Id be happy to know whether the new RC works for you. I didnt notice any memory overhead anymore.

**Comment by brahaney** [comment]
It's fixed! Amazing. Thanks for the speedy response, we were really excited to use this mod! Feel free to ping me if you've got new versions or features to test.

**Comment by Mydayyy** [comment]
Glad to hear!

I released 1.2.1 as a new version on NexusMods and Thunderstore. If you're already running the rc1-1.2.1 there's no need to update, there weren't any changes except a formal release. Incredible sorry for this bug, thank you for the great report!

Cheers
Mydayyy
